File: mini-jdbc/src/main/java/com/mini/core/jdbc/OracleJdbcTemplate.java
package com.mini.core.jdbc;

import javax.sql.DataSource;

import static java.lang.String.format;

public class OracleJdbcTemplate extends JdbcTemplate {
    public OracleJdbcTemplate() {
    }

    public OracleJdbcTemplate(DataSource dataSource) {
        super(dataSource);
    }

    public OracleJdbcTemplate(DataSource dataSource, boolean lazyInit) {
        super(dataSource, lazyInit);
    }

    @Override
    @SuppressWarnings("SpellCheckingInspection")
    protected String paging(int start, int limit, String sql) {
        return format("" +
                "SELECT MAX_COUNT_ROWNUM.* FROM (" +
                "    SELECT MAX_COUNT.*, rownum ROW_NUMBER FROM (" +
                "        %s" +
                "    ) MAX_COUNT WHERE rownum <= %d" +
                ") MAX_COUNT_ROWNUM WHERE ROW_NUMBER > %d" +
                "", sql, start + limit, limit);
    }

    @Override
    protected String totals(String sql) {
        return format("SELECT COUNT(*) FROM (%s) TB", sql);
    }
}

File: mini-jdbc/src/main/kotlin/com/mini/core/jdbc/OracleJdbcTemplate.kt
@file:Suppress("unused")

package com.mini.core.jdbc;

import javax.sql.DataSource

class OracleJdbcTemplate @JvmOverloads
constructor(
        dataSource: DataSource,
        lazyInit: Boolean = true
) : JdbcTemplate(dataSource, lazyInit) {
    override fun paging(start: Int, limit: Int, sql: String): String {
        return """ 
            |SELECT MAX_COUNT_ROWNUM.* FROM ( 
            |    SELECT MAX_COUNT.*, rownum ROW_NUMBER FROM (
            |        $sql
            |    ) MAX_COUNT WHERE rownum <= ${start + limit}"  
            |) MAX_COUNT_ROWNUM WHERE ROW_NUMBER > $limit, 
        """.trimIndent()
    }

    override fun totals(sql: String): String {
        return "SELECT COUNT(*) FROM (${sql}) TB"
    }
}

